File: voice-relay/voice_relay/ingress/mcp.py
"""MCP ingress — Claude / ChatGPT voice.

Two modes, chosen by config:

- **OAuth** (a Google client is configured): the relay is a self-hosted
  OAuth authorization server (voice_relay.oauth) that federates login to
  Google. Identity comes from the validated access token via
  ``get_access_token()`` — no header parsing, no static token.

- **Static bearer** (default / dormant OAuth): the connector presents a
  voice token; identity resolves via the TokenStore.

`build_mcp_app()` returns the MCP ASGI app to be mounted at the ROOT of the
voice subdomain (the OAuth discovery + endpoints must live at root — see
docs/voice-relay-oauth.md). Returns None if the mcp SDK isn't importable or
construction fails — the relay's HTTP/Siri path never depends on MCP.

NB: no ``from __future__ import annotations`` — FastMCP's @tool() introspects
real annotation objects; PEP 563 strings crash it (fixed once already).
"""

import asyncio
import logging
import uuid

from voice_relay.core import Core
from voice_relay.identity import UnknownToken
from voice_relay.types import VoiceRequest

log = logging.getLogger("voice_relay.ingress.mcp")


def build_mcp_app(core: Core, config):
    """Build the MCP ASGI app (root-mounted), or None if MCP is unavailable."""
    try:
        from mcp.server.fastmcp import FastMCP
        from mcp.server.fastmcp.server import Context
    except ImportError:
        log.warning("mcp SDK not installed; MCP ingress disabled.")
        return None

    try:
        if config.oauth_enabled:
            app = _build_oauth(core, config, FastMCP, Context)
            log.info("MCP ingress: OAuth mode (Google upstream) at root")
            return app
        app = _build_bearer(core, config, FastMCP, Context)
        log.info("MCP ingress: static-bearer mode")
        return app
    except Exception:
        log.exception("MCP app failed to build; disabling MCP. Relay stays up.")
        return None


async def _run(core: Core, config, command: str, *, slack_uid: str = "", token: str = "") -> str:
    """Shared: build a VoiceRequest and hand back the spoken reply."""
    req = VoiceRequest(
        text=(command or "").strip(),
        token=token,
        slack_uid=slack_uid,
        client="mcp",
        request_id=f"mcp-{uuid.uuid4().hex[:12]}",
    )
    try:
        reply = await core.handle(req)
    except UnknownToken:
        return (
            "This connector isn't authorized. Generate a voice token in "
            "honeybot (DM it 'generate my voice token') and put it in the "
            "connector's auth settings."
        )
    return reply.speech


# ---------------------------------------------------------------------------
# OAuth mode
# ---------------------------------------------------------------------------
def _build_oauth(core: Core, config, FastMCP, Context):
    from mcp.server.auth.middleware.auth_context import get_access_token
    from mcp.server.auth.settings import AuthSettings, ClientRegistrationOptions
    from pydantic import AnyHttpUrl
    from starlette.responses import PlainTextResponse

    from voice_relay.oauth import slack
    from voice_relay.oauth.provider import HoneybotOAuthProvider
    from voice_relay.oauth.store import Store

    provider = HoneybotOAuthProvider(
        public_url=config.public_url,
        google_client_id=config.oauth_google_client_id,
        google_client_secret=config.oauth_google_client_secret,
        allowed_domains=config.oauth_allowed_domains,
        slack_uid_resolver=slack.resolver(config.slack_bot_token),
        store=Store(config.oauth_store_path),
    )
    mcp = FastMCP(
        "honeybot",
        auth_server_provider=provider,
        auth=AuthSettings(
            issuer_url=AnyHttpUrl(config.public_url),
            client_registration_options=ClientRegistrationOptions(enabled=True),
            required_scopes=[],
        ),
    )

    @mcp.tool()
    async def ask_honeybot(command: str) -> str:
        """Send a command to honeybot and return its response.

        If honeybot can answer quickly it replies inline; otherwise it
        acknowledges and follows up with the result in your Slack DM.
        """
        token = get_access_token()
        uid = getattr(token, "slack_uid", "") if token else ""
        if not uid:
            return "Your session isn't identity-verified. Sign in again."
        return await _run(core, config, command, slack_uid=uid)

    @mcp.custom_route("/oauth/callback", methods=["GET"])
    async def google_callback(request):
        code = request.query_params.get("code", "")
        state = request.query_params.get("state", "")
        err = request.query_params.get("error", "")
        if err or not code or not state:
            return PlainTextResponse(
                f"Sign-in failed: {err or 'missing code/state'}", status_code=400
            )
        try:
            redirect = await provider.complete_google_callback(code, state)
        except PermissionError as e:
            log.warning("oauth callback denied: %s", e)
            return PlainTextResponse(
                "Sign-in denied: your Google account isn't allowed here.",
                status_code=403,
            )
        except Exception as e:  # noqa: BLE001
            log.exception("oauth callback error")
            return PlainTextResponse(f"Sign-in error: {e}", status_code=400)
        # 302 back to the MCP client's redirect_uri with our code.
        return PlainTextResponse(
            "", status_code=302, headers={"Location": redirect}
        )

    return mcp.streamable_http_app()


# ---------------------------------------------------------------------------
# Static-bearer mode (default when OAuth is dormant)
# ---------------------------------------------------------------------------
def _build_bearer(core: Core, config, FastMCP, Context):
    mcp = FastMCP("honeybot")

    @mcp.tool()
    async def ask_honeybot(command: str, ctx: Context) -> str:
        """Send a command to honeybot and return its response."""
        return await _run(core, config, command, token=_bearer_from_ctx(ctx))

    return mcp.streamable_http_app()


def _bearer_from_ctx(ctx) -> str:
    """Best-effort bearer extraction from the MCP request (bearer mode)."""
    try:
        auth = ctx.request_context.request.headers.get("authorization", "")
    except AttributeError:
        return ""
    value = auth.strip()
    return value[7:].strip() if value.lower().startswith("bearer ") else value

File: voice-relay-ts/src/ingress/mcp.ts
/**
 * MCP ingress — Claude/ChatGPT voice, on the Streamable-HTTP transport.
 *
 * Unified auth: /mcp is always protected by requireBearerAuth with a verifier.
 *   - OAuth mode (Google client configured): verifier = the OAuth provider;
 *     mcpAuthRouter adds /authorize /token /register /.well-known at root and
 *     we add the /oauth/callback route. Identity = the validated token.
 *   - Static-bearer mode: verifier resolves the voice token via the TokenStore.
 * Either way the tool reads slackUid from extra.authInfo.extra, so the tool
 * logic is identical.
 */

import { randomUUID } from "node:crypto";
import type { Express, Request, Response } from "express";
import { z } from "zod";
import { McpServer } from "@modelcontextprotocol/sdk/server/mcp.js";
import { StreamableHTTPServerTransport } from "@modelcontextprotocol/sdk/server/streamableHttp.js";
import { isInitializeRequest } from "@modelcontextprotocol/sdk/types.js";
import { requireBearerAuth } from "@modelcontextprotocol/sdk/server/auth/middleware/bearerAuth.js";
import { mcpAuthRouter } from "@modelcontextprotocol/sdk/server/auth/router.js";
import { InvalidTokenError } from "@modelcontextprotocol/sdk/server/auth/errors.js";
import type { AuthInfo } from "@modelcontextprotocol/sdk/server/auth/types.js";
import type { OAuthTokenVerifier } from "@modelcontextprotocol/sdk/server/auth/provider.js";

import { oauthEnabled, type Config } from "../config.js";
import type { Core } from "../core.js";
import { ForbiddenError, HoneybotOAuthProvider } from "../oauth/provider.js";
import { OAuthStore } from "../oauth/store.js";
import { slackUidResolver } from "../slackClient.js";
import type { TokenStore } from "../tokenStore.js";

export function mountMcp(app: Express, deps: { core: Core; tokenStore: TokenStore; config: Config }): void {
  const { core, tokenStore, config } = deps;

  const verifier = oauthEnabled(config)
    ? mountOAuth(app, config)
    : bearerVerifier(tokenStore);

  const auth = requireBearerAuth({ verifier });
  const transports: Record<string, StreamableHTTPServerTransport> = {};

  app.post("/mcp", auth, async (req: Request, res: Response) => {
    const sid = req.headers["mcp-session-id"] as string | undefined;
    let transport = sid ? transports[sid] : undefined;
    if (!transport) {
      if (sid || !isInitializeRequest(req.body)) {
        res
          .status(400)
          .json({ jsonrpc: "2.0", error: { code: -32000, message: "No valid session" }, id: null });
        return;
      }
      transport = new StreamableHTTPServerTransport({
        sessionIdGenerator: () => randomUUID(),
        onsessioninitialized: (id: string) => {
          transports[id] = transport!;
        },
      });
      transport.onclose = () => {
        if (transport!.sessionId) delete transports[transport!.sessionId];
      };
      await buildMcpServer(core).connect(transport);
    }
    await transport.handleRequest(req, res, req.body);
  });

  const sessionRequest = async (req: Request, res: Response): Promise<void> => {
    const sid = req.headers["mcp-session-id"] as string | undefined;
    const transport = sid ? transports[sid] : undefined;
    if (!transport) {
      res.status(400).send("No valid session");
      return;
    }
    await transport.handleRequest(req, res);
  };
  app.get("/mcp", auth, sessionRequest);
  app.delete("/mcp", auth, sessionRequest);
}

function mountOAuth(app: Express, config: Config): OAuthTokenVerifier {
  const provider = new HoneybotOAuthProvider({
    publicUrl: config.publicUrl,
    googleClientId: config.oauthGoogleClientId,
    googleClientSecret: config.oauthGoogleClientSecret,
    allowedDomains: config.oauthAllowedDomains,
    resolveUid: slackUidResolver(config.slackBotToken),
    store: new OAuthStore(config.oauthStorePath),
  });

  app.use(mcpAuthRouter({ provider, issuerUrl: new URL(config.publicUrl) }));

  app.get("/oauth/callback", async (req: Request, res: Response) => {
    const code = String(req.query["code"] ?? "");
    const state = String(req.query["state"] ?? "");
    const err = String(req.query["error"] ?? "");
    if (err || !code || !state) {
      res.status(400).type("text/plain").send(`Sign-in failed: ${err || "missing code/state"}`);
      return;
    }
    try {
      res.redirect(await provider.completeGoogleCallback(code, state));
    } catch (e) {
      if (e instanceof ForbiddenError) {
        res.status(403).type("text/plain").send("Sign-in denied: your Google account isn't allowed here.");
        return;
      }
      res.status(400).type("text/plain").send("Sign-in error.");
    }
  });

  console.log("MCP: OAuth mode (Google upstream)");
  return provider;
}

function bearerVerifier(tokenStore: TokenStore): OAuthTokenVerifier {
  console.log("MCP: static-bearer mode");
  return {
    async verifyAccessToken(token: string): Promise<AuthInfo> {
      try {
        const slackUid = tokenStore.resolve(token);
        return { token, clientId: "voice-token", scopes: [], extra: { slackUid } };
      } catch {
        throw new InvalidTokenError("unknown voice token");
      }
    },
  };
}

function buildMcpServer(core: Core): McpServer {
  const server = new McpServer({ name: "honeybot", version: "0.1.0" });
  server.registerTool(
    "ask_honeybot",
    {
      description:
        "Send a command to honeybot and get its response. Fast answers come " +
        "back inline; slower ones are acknowledged and followed up in your Slack DM.",
      inputSchema: { command: z.string() },
    },
    async ({ command }, extra) => {
      const slackUid = (extra.authInfo?.extra?.["slackUid"] as string | undefined) ?? "";
      if (!slackUid) {
        return { content: [{ type: "text" as const, text: "Your session isn't identity-verified. Sign in again." }] };
      }
      const reply = await core.handle({
        text: (command ?? "").trim(),
        token: "",
        slackUid,
        client: "mcp",
        requestId: `mcp-${randomUUID().slice(0, 12)}`,
      });
      return { content: [{ type: "text" as const, text: reply.speech }] };
    },
  );
  return server;
}
